### ISSUE TITLE
Any suggestion to accelerate parameter update on PS for distributed training?

### ISSUE THREAD

**Issue by ymjiang** [apache/incubator-mxnet#14485]
When doing distributed training, I find the parameter update (push requests) dominates the communication time. Since the update happens on CPU, I wonder if there is any suggestion to accelerate the CPU performance on parameter update?

For simplicity, I use 1 worker (one GPU) and 1 PS, and locate them on different machines (no other task run on them). I use `kvstore=dist_sync` and train Alexnet. Sending one of the key (data size=2MB) only takes about 200us while updating it takes about 780us. 

The update time is measured as follows:
- Before [sending the update request to the engine](https://github.com/apache/incubator-mxnet/blob/master/src/kvstore/kvstore_dist_server.h#L353), get `timestamp_1`
- After `stored.WaitToRead();`, get `timestamp_2`
- Update_time=`timestamp_2-timestamp_1`

From `top` I observe PS only uses about 10 cores (~976% CPU). However, the total number of cores on PS's machine is 64. Is there any option or env that can enable PS to use as more cores as possible to boost performance?

**Comment by apeforest** [comment]
@ymjiang how many parameters in one update

**Comment by ymjiang** [comment]
@apeforest I use Alexnet, about 50M parameter.

**Comment by zachgk** [comment]
Thank you for submitting the issue! I'm labeling it so the MXNet community members can help resolve it.

@mxnet-label-bot add [KVStore, Performance]

**Comment by apeforest** [comment]
@ymjiang There are a few options I suggest you to try:

1) Set the env variable MXNET_KVSTORE_REDUCTION_NTHREADS. This should specify more CPU threads to perfom reduction on hour parameter server
https://mxnet.incubator.apache.org/versions/master/faq/distributed_training.html#environment-variables

2) If the parameter server is the computation bottleneck, try update_on_kvstore=False in your gluon Trainer:
```
trainer = gluon.Trainer(net.collect_params(), optimizer='sgd',
                        optimizer_params={'learning_rate': opt.lr,
                                          'wd': opt.wd,
                                          'momentum': opt.momentum,
                                          'multi_precision': True},
                        kvstore=kv,
                        update_on_kvstore=True)
```
With this, the parameter server only aggregates the gradients, and its the worker that will update weights locally with the gradients.

**Comment by ymjiang** [comment]
@apeforest Thank you very much for your suggestions! I will try `MXNET_KVSTORE_REDUCTION_NTHREADS` to see if it works. 

In terms of option-2, I am using MXNet-Module instead of gluon trainer. So is there any option similar to `update_on_kvstore=False` in Module?

**Comment by ymjiang** [comment]
Ping @eric-haibin-lin

**Comment by szha** [comment]
@ymjiang sorry for the late reply. `update_on_kvstore` is an option in Gluon’s Trainer that it uses to decide whether parameters should be bound to gradient tensors, and if so it would directly apply the gradient on the parameter.

**Comment by ymjiang** [comment]
Hey @szha , thank you very much for the reply. Later I found out that the bottleneck is actually at somewhere else, rather than on PS update. So I closed this issue. BTW, `update_on_kvstore` is an useful information to me :)

**Comment by eric-haibin-lin** [comment]
Sorry for the late reply. On kvstore performance, one potential improvement I find is that there is one extra copy in kvstore-server during pull: https://github.com/apache/incubator-mxnet/blob/master/src/kvstore/kvstore_dist_server.h#L598 

The reason is that the ps-lite response API does not provide an option for callback. If that's added, we can avoid this copy operation and schedule the response using engine->push().

**Comment by ymjiang** [comment]
Hi @eric-haibin-lin , thank you for the note of extra pull copy. Actually I solve it by directly casting the pointer to response vals, instead of using the `CopyFrom`:

`response->vals = ps::SArray<char>(static_cast<char*>(stored.data().dptr_), len, false);`

The performance indeed increases noticeably without the memory copy. And with proper `WaitToRead` I believe this will not affect the  correctness (I have confirmed the convergence with ImageNet dataset).

**Comment by eric-haibin-lin** [comment]
@ymjiang what's your setup and how much improvement did you see?

**Comment by ymjiang** [comment]
@eric-haibin-lin For a very simplistic case with 1 worker (one 1080Ti GPU) & 1 PS and train VGG16 (batch size=32), I observe about 15-20% training speed improvement.
